File: OA/269-alien-dictionary.java
import java.security.DrbgParameters.Reseed;
import java.util.*;

class Solution {
    public String alienOrder(String[] words) {
        Map<Character, Set<Character>> graph = new HashMap<>();
        int[] inDegree = new int[26];
        buildGraph(graph, inDegree, words);
        return bfs(graph, inDegree);
    }

    private void buildGraph(Map<Character, Set<Character>> g, int[] inDegree, String[] words) {
        for (String s : words) {
            for (char c : s.toCharArray()) {
                g.putIfAbsent(c, new HashSet<>());
            }
        }

        for (int i = 1; i < words.length; i++) {
            String first = words[i - 1];
            String second = words[i];
            int len = Math.min(first.length(), second.length());

            for (int j = 0; j < len; j++) {
                if (first.charAt(j) != second.charAt(j)) {

                    char out = first.charAt(j);
                    char in = second.charAt(j);

                    if (!g.get(out).contains(in)) {
                        g.get(out).add(in);
                        inDegree[in - 'a']++;
                    }
                    break;
                }

            }
        }
    }

    private String bfs(Map<Character, Set<Character>> g, int[] inDegree) {
        StringBuilder sb = new StringBuilder();
        int totalChars = g.size();
        Queue<Character> q = new LinkedList<>();
        for (char c : g.keySet()) {
            if (inDegree[c - 'a'] == 0) {
                sb.append(c);
                q.offer(c);
            }
        }

        while (!q.isEmpty()) {
            char cur = q.poll();
            if (g.get(cur) == null || g.get(cur).size() == 0)
                continue;

            for (char neigh : g.get(cur)) {
                inDegree[neigh - 'a']--;
                if (inDegree[neigh - 'a'] == 0) {
                    q.offer(neigh);
                    sb.append(neigh);
                }
            }
        }

        return sb.length() == totalChars ? sb.toString() : "";
    }
}

class Main {
    public static void main(String[] args) {
        System.out.println("Hello world!");
        Solution solution = new Solution();
        String[] words = new String[] { "zy", "zx" };

        System.out.println(solution.alienOrder(words));
    }
}

File: 269-alien-dictionary.js
/**
 * @param {string[]} words
 * @return {string}
 */

/**
 * @param {string[]} words
 * @return {string}
 */

class Graph {
  constructor() {
    this.graph = new Map();
    this.inDegree = new Map();
  }

  buildGraph = words => {
    // Add vertices and initialize inDegree
    for (let word of words) {
      for (let c of word) {
        this.graph.set(c, new Set());
        this.inDegree.set(c, 0);
      }
    }
    // Add edges and update indegree for each
    for (let i = 1; i < words.length; i++) {
      const first = words[i - 1],
        second = words[i];
      let len = Math.min(first.length, second.length);

      for (let j = 0; j < len; j++) {
        if (first.charAt(j) != second.charAt(j)) {
          let u = first.charAt(j),
            v = second.charAt(j);

          if (!this.graph.get(u).has(v)) {
            this.graph.get(u).add(v);
            this.inDegree.set(v, this.inDegree.get(v) + 1);
          }
          break;
        }
      }
    }
  };

  topologicalSort() {
    let q = [];
    for (let c of this.graph.keys()) {
      if (this.inDegree.get(c) == 0) q.push(c);
    }

    let str = "";
    while (q.length != 0) {
      let c = q.shift();
      str += c;

      for (let neighbor of this.graph.get(c)) {
        this.inDegree.set(neighbor, this.inDegree.get(neighbor) - 1);
        if (this.inDegree.get(neighbor) == 0) q.push(neighbor);
      }
    }
    return str.length == this.graph.size ? str : "";
  }
}

var alienOrder = function(words) {
  const graph = new Graph();
  graph.buildGraph(words);
  return graph.topologicalSort();
};
